Extract PubMed IDs from PubMed URLs

mudge opened the pull request:
Support the extraction of PubMed IDs from PubMed record page URLs such as https://www.ncbi.nlm.nih.gov/pubmed/123 and https://www.ncbi.nlm.nih.gov/m/pubmed/123/

As with other PubMed IDs, strip any leading zeroes from these IDs (e.g.  https://www.ncbi.nlm.nih.gov/m/pubmed/000123/ will return 123, not 000123).

mudge committed:
Test that each identifier gracefully handles null

So that users can pass null to any extract method without throwing an
error, ensure that this behaviour is encoded in the tests.

mudge committed:
Extract PubMed IDs from PubMed URLs

Support the extraction of PubMed IDs from PubMed record page URLs such
as https://www.ncbi.nlm.nih.gov/pubmed/123 and
https://www.ncbi.nlm.nih.gov/m/pubmed/123/

As with other PubMed IDs, strip any leading zeroes from these IDs (e.g.
https://www.ncbi.nlm.nih.gov/m/pubmed/000123/ will return 123, not
000123).

mudge commented on src/PubmedId.php:
```diff
@@ -3,9 +3,38 @@
+        |
+        \#  # Fragment
+    )
+}xu
```
This regexp is in contrast to @jbilbo's angle of attack in https://github.com/altmetric/identifiers/pull/20/ which is to have _two_ separate patterns.

At first (when doing the prefixes), I thought this might be simpler but the number of required suffixes might change my mind.

jbilbo replied:
The problem is the required suffixes for a zero-padded-integer pattern are different from the PubmedURL pattern, so the result code is not equivalent as having both separate, e.g. I don't think the zero-padded-integer should have `/`, `?` or `#` as acceptable suffixes, in the current regexp we'll extract `123456` from `123456#ABC` which looks wrong to me.

mudge replied:
That makes sense: I'll add some tests to ensure that doesn't happen and split the two patterns up.

mudge committed:
Support URIs with PubMed schemes

Add support for extracting PubMed IDs out of URIs using the pmid or
PubMed info scheme.

mudge committed:
Separate PubMed URI and ID extraction

As the form of standalone PubMed IDs differs sufficiently from PubMed
IDs within URIs (e.g. pmid:12345 or info:pmid/12345), split up the two
patterns and match against them separately.
